### PR TITLE
RFP-10 milestone 3

### PR DESCRIPTION
**Pull request by chappjc**
Requires unmerged dcrrpcclient PR: https://github.com/decred/dcrrpcclient/pull/35 (and an updated glide.yaml/glide.lock).

Milestone 3 of RFP-10 includes the following legacy RPC tests:
- [x] gettickets,
- [x] purchaseticket
- [x] get/setticketmaxprice
- [x] get/setbalancetomaintain
- [x] getstakeinfo
- [x] walletinfo

Note that there are no getticketmaxprice or getbalancetomaintain commands and as such walletinfo is used to test the mutators.

### CODE REVIEW

**Comment by chappjc** [comment]
This completes RFP-10 milestone 3 and RFP-10.

Tests added or completed:

gettickets, purchaseticket, get/setticketmaxprice, get/setbalancetomaintain, getstakeinfo, walletinfo

Note that there are no getticketmaxprice or getbalancetomaintain commands. As such, walletinfo is used to test the corresponding mutators.

Also note that the addition of the following functions to dcrrpcclient package were necessary:  SetTicketMaxPrice (https://github.com/decred/dcrrpcclient/pull/34), and SetBalanceToMaintain (https://github.com/decred/dcrrpcclient/pull/35).

Other fixes/updates included:

- Following rebase with master, switch from using OrganizationAddress to OrganizationPkScript with ExtractPkScriptAddrs.
- Change IsSSTx call, which now needs a wire.MsgTx rather than Tx as before.
- Handle errors in teardown of harnesses better by not skipping remaining harnesses if one fails to tear down.
- Swap out problematic utxo check using ListUnspent with the one that used GetTxOut taking into account the previous outpoint index.
- Add helper advanceToNewWindow to generate blocks until the start of a new ticket price window.  This is defined as the point at which next height difficulty != current height difficulty.
- Add helper advanceToHeight to generate new blocks until the desired block height is reached.  Use this with care or the chain may stall if insufficient mature tickets are available for block validation.
- Add helper getBestBlockHeight, which only returns height.
- Add helper mustGetStakeInfo, which returns only a dcrjson.GetStakeInfoResult or calls t.Fatal on error.
- Remove all hard-coded spend limit for ticket purchases.  Use stake diff of next block to set price limit.
- Add helpers mustGetStakeDiff and mustGetStakeDiffNext, which returns only the stake difficulty at the current and next blocks, respectively, or call t.Fatal on error.
- Add function to advance to given block height by generating blocks.  Use this at the start of tests that require stake validation height (voting started).
- Introduce more ticket purchases and increase the number of tickets purchased where possible to make sure chain doesn't stall.
- Add helper includesStakeTx, which checks if a block contains a given transaction hash in it's stake transactions (STransactions).

Any remaining comments with "TODO" are not urgent or necessary, but should be considered in the future.  Some topics for further consideration (beyond RFP-10):

- How can the wallet become resynced more quickly after the node restarts?  It takes almost a minute of waiting, but no actual work being done.  https://github.com/decred/dcrwallet/issues/364
- Perhaps it's best to redo the harness entirely and use a wallet with a pre-determined seed, and thus mining address?
- How best to handle (in test and in app) resend of votes causing an error, as when SetGenerate(true, 0) is run when the wallet's votes were already cast or the block.  Ignore all errors from SetGenerate or detect this specific error to ignore it?
- The address field of ListUnspent results is not always useful as it only includes a single address when there may be many.  This necessitates extra calls to txscript.ExtractPkScriptAddrs.  https://github.com/decred/dcrwallet/issues/370
- Perhaps a wallet in stakepool mode can be used to test some of the pool-specific functionality, such as for purchaseticket.
- What other functions are missing from dcrrpcclient?

**Comment by alexlyp** [comment]
Can you update glide.lock so it only updates dcrd and dcrrpclient?  

We're trying to avoid updating 3rd party deps in PRs that are unrelated.

**Comment by chappjc** [comment]
No problem.  I'll update tonight.

**Comment by alexlyp** [comment]
First pass looks good.  Only took a short 520s to finish.  :)  
Anything you think we can do to get that down some?

**Comment by chappjc** [comment]
The Sleeps can probably be reduced with 0.5.0 speed improvements to block processing.
But the delay setting up the harnesses is an issue I reported, and I haven't been able to pin down.

**Comment by chappjc** [comment]
TestRpcServer is down to 120 sec.  Could probably be less, but the long wait during harness setup is the big delay now at 220 sec, and it should be 20 sec.

**Comment by alexlyp** [comment]
OK runs a lot faster now good job on the speed up.

Just ran through a review and looks mostly clean.  
I noted a bunch of lingering spots with commented out code that should be turned into a comment or removed entirely.  If it is a TODO for later note as such.

I also noted a bunch of times you skipped the err from NewAmount.  I'd prefer it if you caught these since (though unlikely) it would cause a panic when trying to use the nil created amount.

**Comment by chappjc** [comment]
@ay-p, @jy-p Is everything A-OK?

**Comment by alexlyp** [comment]
There were still a few lingering things I noted in the review.  Tx --> tx, ii --> i another skipped err

After that should be good to go.

**Comment by chappjc** [comment]
Done.  The ii thing is habit for me as i is imaginary i in MATLAB and best to avoid.  Obviously not with go.  rebasing...

**Comment by chappjc** [comment]
@alexlyp I'm looking at your comments, but it looks like you reviewed a previous version.

File: rpctest/rpcharness_test.go [hunk 1]
```diff
@@ -1608,20 +1634,21 @@ func testGetSetTicketFee(r *Harness, t *testing.T) {
...
+	priceLimit, _ := dcrutil.NewAmount(2 * mustGetStakeDiffNext(r, t))
+	hashes, err := wcl.PurchaseTicket("default", priceLimit,
 		&minConf, nil, &numTicket, nil, nil, nil)
 	if err != nil {
-		t.Fatal("Unable to purchase with nil ticketAddr:", err)
+		t.Fatal("Unable to purchase ticket:", err)
 	}
-	if len(hashes) != 1 {
+	if len(hashes) != numTicket {
 		t.Fatal("More than one tx hash returned. Expected one.")
```
Comment (chappjc) [inline rpctest/rpcharness_test.go]: Ah, good catch.

File: rpctest/rpcharness_test.go [hunk 2]
```diff
@@ -1655,6 +1682,91 @@ func testGetSetTicketFee(r *Harness, t *testing.T) {
...
+	ticketHashes, err = wcl.GetTickets(true)
+	if err != nil {
+		t.Fatal("GetTickets failed:", err)
+	}
+
+	numTicketsInit := len(ticketHashes)
+
+	// Purchase a full blocks worth of tickets
+	minConf, numTicketsPurchased := 1, int(chaincfg.SimNetParams.MaxFreshStakePerBlock)
+	priceLimit, _ := dcrutil.NewAmount(2 * mustGetStakeDiffNext(r, t))
```
Comment (chappjc) [inline rpctest/rpcharness_test.go]: Huh, missed this one.

**Comment by chappjc** [comment]
@alexlyp I think your review wasn't actually submitted until today.  I'll have go back to make sure I took your advice as expected

File: rpctest/rpcharness_test.go (hunk 2 above)
Comment (chappjc) [inline rpctest/rpcharness_test.go]: This was fixed.  I think the code here is outdated.

File: rpctest/rpcharness_test.go [hunk 3]
```diff
@@ -1655,6 +1682,91 @@ func testGetSetTicketFee(r *Harness, t *testing.T) {
...
+	ticketHashes, err = wcl.GetTickets(true)
+	if err != nil {
+		t.Fatal("GetTickets failed:", err)
+	}
+
+	if numTicketsInit+numTicketsPurchased != len(ticketHashes) {
+		t.Fatal("GetTickets(true) did not include unmined tickets")
+	}
+
+	// Ensure that there are more total tickets than live tickets
```
Comment (chappjc) [inline rpctest/rpcharness_test.go]: discussed in IRC.  Addressed with comments.

**Pull request merged**